src/plugins/intel_npu/src/backend/src/zero_infer_request.cpp

```diff
@@ -284,18 +300,119 @@ ZeroInferRequest::ZeroInferRequest(const std::shared_ptr<ZeroInitStructsHolder>&
+                } else {
+                    allocator = zeroMemory::HostMemAllocator(_initStructs);
+                };
+
+                _copyAllTensors.at(name) = ov::make_tensor(tensor->get_element_type(), tensor->get_shape(), allocator);
+                if (isState) {
+                    _copyAllTensors.at(stateInputName) = _copyAllTensors.at(name);
+                }
+
+                _tensorsData.at(name) = TensorData{copyTensor->data(), copyTensor->get_byte_size()};
+                if (isState) {
+                    _tensorsData.at(stateInputName) = TensorData{copyTensor->data(), copyTensor->get_byte_size()};
+                }
 
-    _logger.debug("ZeroInferRequest::ZeroInferRequest - constructing pipeline");
-    /// Construct pipepline
-    _pipeline = makePipeline(_executorPtr,
-                             _config,
-                             _profilingPool,
-                             _profilingQuery,
-                             _npuProfiling,
-                             _copyAllTensors,
-                             _batchSize);
-    _logger.debug("ZeroInferRequest::ZeroInferRequest - SyncInferRequest completed");
+                _createPipeline = true;
+            }
+        } else {
+            ze_context_handle_t l0_context = nullptr;
+            void* data = nullptr;
+
+            auto context_properties = remoteTensor->get_context()->get_property();
+            for (auto&& value : context_properties) {
+                if (value.first == ov::intel_npu::l0_context.name()) {
+                    l0_context = reinterpret_cast<ze_context_handle_t>(value.second.as<void*>());
+                    break;
+                }
+            }
+
+            if (_initStructs->getContext() != l0_context) {
+                OPENVINO_THROW("Using different context for creating the tensor is not supported");
+            }
+
+            auto tensor_properties = remoteTensor->get_properties();
+            for (auto&& value : tensor_properties) {
+                if (value.first == ov::intel_npu::mem_handle.name()) {
+                    data = value.second.as<void*>();
+                    break;
+                }
+            }
+
+            if (data == nullptr) {
+                OPENVINO_THROW("Empty buffer");
+            }
+
+            auto& copyTensor = _copyAllTensors.at(name);
+            auto remoteCopyTensor = std::dynamic_pointer_cast<ZeroRemoteTensor>(copyTensor);
+            if (remoteCopyTensor == nullptr) {
+                _copyAllTensors.at(name) = tensor;
+                if (isState) {
+                    _copyAllTensors.at(stateInputName) = _copyAllTensors.at(name);
+                }
+
+                _tensorsData.at(name) = TensorData{data, remoteTensor->get_byte_size()};
+                if (isState) {
+                    _tensorsData.at(stateInputName) = TensorData{data, remoteTensor->get_byte_size()};
+                }
+
+                _createPipeline = true;
```

Comment:
This block is the same on both sides of the branch. I suggest hoisting it out to simplify the flow.